Full documentation of AZSL & AZSLc. (#1158)
o3de/o3de.org: content/docs/atom-guide/dev-guide/shaders/azsl/shader-variant-options.md

---
linktitle: Shader Variant Options

> title: AZSL, Design Principles And Purpose Of Shader Variant Options

Reviewer: oxford comma needed

> title: AZSL, Design Principles And Purpose Of Shader Variant Options

Author: Fixed.

> description: Learn about the design principles and purpose of AZSL Shader Variant Options in the Atom Renderer. 

Author: Fixed.

description: Learn about the design principles and purpose of AZSL Shader Variant Options in the Atom Renderer. 
toc: true
weight: 300

> draft: true

Author: What does it mean to publish hidden? To me it's just another regular doc.

> draft: true

Reviewer 2: This may have been set as "draft" previously, because it didn't have content. 
If it's `draft: true`, then it won't publish on the website. You are right to delete this line.

> Shader Variant Options are "statically optimizable" Shader Constants that the developer can choose to compile as static constants or as regular global variables. They are only used in conditional statements.  

Author: Fixed.

> Shader Variant Options are "statically optimizable" Shader Constants that the developer can choose to compile as static constants or as regular global variables. They are only used in conditional statements.  

Reviewer 2: @galibzon  Can you also lower case "shader variant options" in the `/azsl/_index.md` file? For consistency with terminology.

> Shader Variant Options are conceptually equivalent to [Specialization Constants](https://www.khronos.org/registry/vulkan/specs/1.1-khr-extensions/html/chap10.html#pipelines-specialization-constants) (external link) in Vulkan.  

Author: In my experience, most people struggle to understand what are shader variant options, and pointing to existing equivalents always helps.

  
The idea is easier to explain with some code. Consider the following AZSL code snippet, where a Shader Variant Option is being used to conditionally branch:  
  
    if (o_useRed) {
        color = float3(1, 0, 0);
    } else {
        color = float3(0, 0, 1);

>     }

Author: Fixed.

  
When the shader code, shown above, is compiled and the value of `o_useRed` is not known at compilation time then `o_useRed` will be compiled as a regular Shader Constant. The shader byte code **will branch at runtime** depending on the value of `o_useRed` as defined by the Application.  
  

> Please notice that **"will branch at runtime"** was emphasized. In general, branching is slower than not branching. To avoid branching, the value of `o_useRed` can be defined at compilation time, let's say to `true`.  

Author: Fixed.

  
    // Because the developer chose o_useRed to be true at compilation time,
    // the compiler will optimize and produce the following branch-less code.

>     color = float3(1, 0, 0);

Author: fixed.

  
Alternatively, at compilation time, `o_useRed` may be defined as `false`, and the optimized output code will look like:  
  
    // Because the developer chose o_useRed to be false at compilation time,
    // the compiler will optimize and produce the following branch-less code.

>     color = float3(0, 0, 1);

Author: Fixed.

  
We can see from this simple example that there can be three different versions of the compiled shader code to be used at runtime. These are called Shader Variants (**ShaderVariantAssets**), and they only differ only in two ways:  
1. Whether the value of some Shader Variant Options is statically defined at compilation time or not.

> 2. And the values chosen at compilation time for those statically defined Shader Variant Options.

Author: Fixed.

  
Following up with this example, We can endup with three different versions (Variants) of the shader bytecode:  
1. The first variant is the one where `o_useRed` is just a regular Shader Constant. This byte code will branch at runtime, BUT provides flexibility to the application to set its value before the shader is submitted for execution. This is called the Root Variant (Root ShaderVariantAsset).
2. The second variant is the one where `o_useRed` was given the value `true` at compilation time. This byte code won't branch at runtime, there are APIs to help the Application discover this variant of the shader before submitting for execution.

> 3. The third variant is the one where `o_useRed` was given the value `false` at compilation time. This byte code won't branch at runtime.  

Author: Fixed

> Using Baked Shader Variants will provide the maximum runtime performance of shaders, at the expense of more shader byte code files in the asset catalog (more ShaderVariantAssets).  

Author: Fixed.

> Depending on the amount of Shader Variant Options declared in a shader, the application may end up with a large permutation of shader variant assets. More on this topic [here](shader-variants-fallback-key).  

Reviewer 3: Feels like this needs to be highlighted with more explanation.

```suggestion
{{< important >}}
If many shader variant options are declared in a shader, a large number of shader variant assets might be generated. This is particularly important if the shader code has nested conditional statements that have shader variant options. A shader variant asset is generated for each possible combination shader variant options, which creates an exponential number of shader variant assets.

For more information on this topic refer to [Shader Variant Options and the Fallback Key](../shader-variants-fallback-key).
{{< /important >}}
```

Author: Fixed.

---

> linktitle: Shader variant options
> title: Shader variant options in Amazon Shading Language (AZSL)

Reviewer 2: ```suggestion
linktitle: Shader Variant Options
title: Shader Variant Options in Amazon Shading Language (AZSL)
```
* Use camel casing for titles. For example: Shader **V**ariant **O**ptions

Author: Fixed.

---
linktitle: Shader variant options
title: Shader variant options in Amazon Shading Language (AZSL)
description: Learn about Amazon Shading Language (AZSL) shader variant options for Atom Renderer. 
weight: 100
---

Shader variant options are shader constants that are only used in conditional statements that can be statically optimized. You can choose to compile them as static constants or as global variables.  
  
Shader variant options are conceptually equivalent to [Specialization Constants](https://www.khronos.org/registry/vulkan/specs/1.1-khr-extensions/html/chap10.html#pipelines-specialization-constants) (external link) in Vulkan.  
  

> The idea is easier to explain with some code. Consider the following AZSL code snippet, where a Shader Variant Option is being used to conditionally branch:  

Author: Removed the link... But I really lkied it :-)

```cpp
    if (o_useRed) {
        color = float3(1, 0, 0);
    } else {
        color = float3(0, 0, 1);
    }
```

> In the shader code above, the value of `o_useRed` is not known at compilation time, so `o_useRed` is compiled as a shader constant. The shader byte code branches at runtime depending on the runtime value of `o_useRed`.  In general, branching is slow and should be avoided.
> In this scenario, if the value of `o_useRed` is defined as `true` at compilation time, then the compiler produces the following branch-less code.  

Reviewer 2: - Add an empty line between.

Currently, you have two paragraphs separated into two lines "In the shader code…" and "In this scenario...". To render a  space between two paragraphs, create an empty line between the paragraphs. If you don't do this, then "In this scenario..." renders as part of the paragraph before. 

Notice in this example that they are in the same paragraph:
![image](https://user-images.githubusercontent.com/75444793/144668952-f21455c6-cdaa-4047-9d50-1ee7f9db2f37.png)

```suggestion
In the shader code above, the value of `o_useRed` is not known at compilation time, so `o_useRed` is compiled as a shader constant. The shader byte code branches at runtime depending on the runtime value of `o_useRed`.  In general, branching is slow and should be avoided.

In this scenario, if the value of `o_useRed` is defined as `true` at compilation time, then the compiler produces the following branch-less code.  
```

Author: Fixed.

```cpp
    // Because the developer chose o_useRed to be true at compilation time,

>     // the compiler will optimize and produce the following branch-less code.

Reviewer 2: Remove the comments below (`// ...`). You already explain this in the sentence above. 
```suggestion
```

Author: Removed.

    color = float3(1, 0, 0);
```
If `o_useRed` is defined as `false` at compilation time, then the compiler produces the following branch-less code:  
```cpp
    // Because the developer chose o_useRed to be false at compilation time,

>     // the compiler will optimize and produce the following branch-less code.

Reviewer 2: Remove the comments below (`// ...`). You already explain this in the sentence above. 
```suggestion
```

Author: Removed.
